---
title: "API Calls"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{api_calls}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = TRUE
)
```

```{r setup}
library(gpmapr)
```

## API calls

```{r health-check}
# Check API health
result <- health_api()
result$status
```

Returns list of all traits.  
```{r get-all-traits}
all_traits <- traits_api()
first_trait <- all_traits$traits[1, ]
first_trait$trait_name
```

Returns a specific trait.
```{r get-specific-trait}
specific_trait <- trait_api(first_trait$id)
specific_trait$trait$trait_name
nrow(specific_trait$coloc_groups)
nrow(specific_trait$study_extractions)
nrow(specific_trait$rare_results)
```

Returns a specific trait with associations.
```{r get-trait-with-associations}
trait_with_associations <- trait_api(first_trait$id, include_associations = TRUE)
trait_with_associations$trait$trait_name
nrow(trait_with_associations$coloc_groups)
nrow(trait_with_associations$associations)
```

Returns list of all genes.
```{r get-all-genes}
all_genes <- genes_api()
first_gene <- all_genes$genes[1, ]
first_gene
```

Returns a specific gene.
```{r get-specific-gene}
# you can call using 'symbol' or 'id' from the above list
specific_gene <- gene_api(first_gene$gene)
specific_gene$gene$gene
nrow(specific_gene$genes_in_regions)
nrow(specific_gene$coloc_groups)
```

You can also include associations
```{r get-gene-with-associations}
gene_with_associations <- gene_api(first_gene$gene, include_associations = TRUE)
gene_with_associations$gene$gene
nrow(gene_with_associations$genes_in_regions)
nrow(gene_with_associations$coloc_groups)
nrow(gene_with_associations$associations)
```

There are various ways you can get information about a variant.
```
snp_id: 123, as supplied by the API.
rsid: rs123456789
variant: chr:bp_ea_oa, where ea is alphabetically < oa
grange: chr:start_bp-stop_bp
```

Variants calls do not include associations.  To do so, you can call variant()

The following calls obtain informaiton about a list of variants, you can optionally include associations at a p-value threshold.

```{r get-variants}
# you can call using 'snp_ids', 'variants', 'rsids', or 'grange'
snp_ids <- head(gene_with_associations$coloc_groups$snp_id, 5)

variants_by_snp_id_api(snp_ids)
variants_by_grange_api(20, 50948630, 50948640)

```
With all the above you can also include associations, and optionally set a p-value threshold
`variants_by_snp_id(snp_ids, include_associations = TRUE, p_value_threshold = 0.05)`
`variants_by_grange(20, 50948630, 50948640, include_associations = TRUE, p_value_threshold = 1)`

To obtain all colocalizing and rare variant information for a variant, use the variant() call.  This automatically includes associations.
```{r get-variant-info}
variant_info <- variant_api(snp_ids[1])
variant_info$variant$snp
nrow(variant_info$coloc_groups)
nrow(variant_info$associations)
```

Get LD proxies for a variant.
```{r get-ld-proxies}
ld_proxies_by_snp_id_api(snp_ids)
```

Get LD matrix for a variant.
```{r get-ld-matrix}
ld_matrix_by_snp_id_api(snp_ids)
```

